### ISSUE TITLE
[Feature request] Allow plain enum lookup to use both inner (enum) and outer scopes

### ISSUE THREAD

**Issue by kirillbobyrev** [clangd/clangd#931]
For this code

```c++
namespace ns {
enum Color { Green };
enum class Vehicle { Car };
}
```

`Dex::fuzzyFind` will not return any results when querying `ns::Color::` - `Color::Green` will only show up for `ns::` requests.

- `Vehicle::Car` scope is `ns::Vehicle::` - it can't be accessed without spelling full enum class name
- `Color::Green` scope is `ns::` because it can be accessed as `ns::Car`

I've had an [attempt](https://reviews.llvm.org/D113974) to reverse this behavior but it's not an improvement. A real implementation would likely have `clangd::Symbol` allow having multiple scopes (primary `ns::` and secondary `ns::Color::`) and then adding the symbol to the both scopes' posting lists.

**Comment by sam-mccall** [comment]
What's the feature you're trying to enable here?
For code completion, `ns::Color::^` is provided by SemaCodeComplete.

**Comment by kirillbobyrev** [comment]
> What's the feature you're trying to enable here? For code completion, `ns::Color::^` is provided by SemaCodeComplete.

Right, but that's only when `ns::Color` is in the scope, isn't it?

The problem I'm actually having is `workspace/symbols` lookup being empty for `ns::Color::`.

**Comment by kadircet** [comment]
I believe this work as intended for workspace symbols 
![Screenshot from 2021-11-19 13-35-16](https://user-images.githubusercontent.com/897555/142623791-40827bbb-37b2-4737-a490-ec6b0f35d63f.png)

you might be seeing different behavior because of filtering on the index implementation.

---

having these symbols available for codecompletion is a different problem though, as we actually rely on AST to provide symbols inside records (as Sam pointed out) and AST just doesn't do partially qualified name lookups.

**Comment by HighCommander4** [comment]
I think the issue is, if the search query is `ns2::Outer1::`, `Inner` is not offered as a result.

**Comment by sam-mccall** [comment]
OK, so we support both in code complete but not workspaceSymbol.

Currently a symbol in the index can only have a single scope. This affects enums and inline namespaces (and AFAIK the latter is fairly obscure).

Frankly I'm not sure that fixing workspaceSymbol for Color::RED is worth complicating the model if RED and Color both work. WDYT?

**Comment by kirillbobyrev** [comment]
AFAIK this only works for code complete if it's surfaced through Sema. That is, if the enum is already in the translation unit.

Here's an example: I'm opening a random file in clangd and would like to use `llvm::remarks::BlockIDs`. Enum type is not included from anywhere:

<img width="906" alt="Screenshot 2021-11-19 at 21 59 21" src="https://user-images.githubusercontent.com/3352968/142691180-08e76616-24ba-4e02-865b-2f1d6dc6c048.png">

And so I try doing that but I don't get any results.

<img width="666" alt="Screenshot 2021-11-19 at 21 59 28" src="https://user-images.githubusercontent.com/3352968/142691745-d8bd1edf-b9d2-4eda-bc13-ac9f823724d6.png">

Unless I know exactly what I'm looking for it might be hard to get the enum value I need:

<img width="659" alt="Screenshot 2021-11-19 at 21 59 06" src="https://user-images.githubusercontent.com/3352968/142691806-1dc9e055-09a1-4f7d-a2b2-d81fed1efbe3.png">

I frequently use CC to "explore" enum values and sometimes enum types might not be imported or forward declared (I guess this won't work too; plain enums can't be forward declared in C++ but C still allows that - so the type will still be in scope but if the "original" header is not included I'm not in luck), this prevents explorative code completion and is maybe not optimal.

**Comment by sam-mccall** [comment]
Yes, if you keep typing into BlockIDs rather than accepting the completion and the include, then you won't get any results.
This is the same as for classes. Namespaces do work though, we have a hack for that.

I see that would be convenient but I'm not sure how valuable making this work for enums only would be.

**Comment by HighCommander4** [comment]
> Yes, if you keep typing into BlockIDs rather than accepting the completion and the include, then you won't get any results.

I would imagine in a typical scenario, `BlockIDs` would be one entry in a long list of names inside the namespace, so "typing into BlockIDs" is likely needed to make the desired completions show up (without having to scroll down a lot) in the first place.

**Comment by sam-mccall** [comment]
Yes :-( You can *accept* `BlockIDs` instead, and then type `::` to keep going, but this may be awkward.
In some scenarios where the expected-type is available like `setID(llvm::remarks::^)` then ranking should bring the values to the top, but really you want a simple exhaustive list in such a case.

Maybe this is worth doing, it just seems a bit fiddly.
